src/main/java/com/epam/aidial/core/service/ResourceService.java

```diff
@@ -246,44 +300,132 @@ public String getResource(ResourceDescription descriptor, boolean lock) {
 
-            if (result.exists && !overwrite) {
-                return null;
+            if (metadata != null) {
+                if (!overwrite) {
+                    return null;
+                }
+
+                etag.validate(metadata.getEtag());
             }
 
-            long updatedAt = time();
-            long createdAt = result.exists ? result.createdAt : updatedAt;
-            redisPut(redisKey, new Result(body, createdAt, updatedAt, false, true));
+            long updatedAt = System.currentTimeMillis();
+            Long createdAt = metadata == null ? (Long) updatedAt : metadata.getCreatedAt();
+            String newEtag = EtagBuilder.generateEtag(body);
+            Result result = new Result(body, newEtag, createdAt, updatedAt, contentType, (long) body.length, false);
+            if (body.length <= maxSize) {
+                redisPut(redisKey, result);
+                if (metadata == null) {
+                    String blobKey = blobKey(descriptor);
+                    blobPut(blobKey, result.toStub()); // create an empty object for listing
+                }
+            } else {
+                flushToBlobStore(redisKey);
+                String blobKey = blobKey(descriptor);
+                blobPut(blobKey, result);
+            }
 
             ResourceEvent event = new ResourceEvent()
                     .setUrl(descriptor.getUrl())
                     .setAction(ResourceEvent.Action.UPDATE)
                     .setTimestamp(updatedAt);
 
-            if (!result.exists) {
+            if (metadata == null) {
                 event.setAction(ResourceEvent.Action.CREATE);
-                blobPut(blobKey, "", createdAt, updatedAt); // create an empty object for listing
             }
 
             topic.publish(event);
-            return new ResourceItemMetadata(descriptor).setCreatedAt(createdAt).setUpdatedAt(updatedAt);
+            return descriptor.getType() == ResourceType.FILE
+                    ? toFileMetadata(descriptor, result)
+                    : toResourceItemMetadata(descriptor, result);
         }
     }
 
+    public FileMetadata putFile(ResourceDescription descriptor, byte[] body, EtagHeader etag, String contentType) {
+        if (descriptor.getType() != ResourceType.FILE) {
+            throw new IllegalArgumentException("Expected a file, got %s".formatted(descriptor.getType()));
+        }
+
+        return (FileMetadata) putResource(descriptor, body, etag, contentType, true, true);
+    }
+
+    public void completeMultipartUpload(
+            ResourceDescription descriptor, MultipartUpload multipartUpload, List<MultipartPart> parts, EtagHeader etag) {
+        String redisKey = redisKey(descriptor);
+        try (var ignore = lockService.lock(redisKey)) {
+            etag.validate(() -> getEtag(descriptor));
+
+            flushToBlobStore(redisKey);
+            blobStore.completeMultipartUpload(multipartUpload, parts);
+        }
```

Comment:
publish event similar to put (please add test)

Reply:
Added